stmhal/dma.c

```diff
@@ -56,34 +196,70 @@ static const uint8_t dma_irqn[NSTREAM] = {
+const dma_descr_t dma_I2C_1_RX = { DMA1_Channel7, DMA_REQUEST_3, DMA_RX_TRANSFER, dma_id_6  , &dma_init_struct_spi_i2c};
+    // DMA2 streams
+const dma_descr_t dma_SPI_3_RX = { DMA2_Channel1, DMA_REQUEST_3, DMA_RX_TRANSFER, dma_id_7  , &dma_init_struct_spi_i2c};
+const dma_descr_t dma_SPI_3_TX = { DMA2_Channel2, DMA_REQUEST_3, DMA_TX_TRANSFER, dma_id_8  , &dma_init_struct_spi_i2c};
+/* Not prefered
+const dma_descr_t dma_ADC_1_RX = { DMA2_Channel3, DMA_REQUEST_0, DMA_RX_TRANSFER, dma_id_9  , NULL};
+const dma_descr_t dma_SPI_1_RX = { DMA2_Channel3, DMA_REQUEST_4, DMA_RX_TRANSFER, dma_id_9  , &dma_init_struct_spi_i2c};
+const dma_descr_t dma_ADC_2_RX = { DMA2_Channel4, DMA_REQUEST_0, DMA_RX_TRANSFER, dma_id_10 , NULL};
+const dma_descr_t dma_DAC_1_TX = { DMA2_Channel4, DMA_REQUEST_3, DMA_TX_TRANSFER, dma_id_10 , NULL};
+const dma_descr_t dma_SPI_1_TX = { DMA2_Channel4, DMA_REQUEST_4, DMA_TX_TRANSFER, dma_id_10 , &dma_init_struct_spi_i2c}; */
+#if MICROPY_HW_HAS_SDCARD
+const dma_descr_t dma_SDIO_0_TX= { DMA2_Channel4, DMA_REQUEST_7, DMA_TX_TRANSFER, dma_id_10 , &dma_init_struct_sdio};
+#endif
+/* Not prefered
+const dma_descr_t dma_ADC_3_RX = { DMA2_Channel5, DMA_REQUEST_0, DMA_RX_TRANSFER, dma_id_11 , NULL};
+const dma_descr_t dma_DAC_2_TX = { DMA2_Channel5, DMA_REQUEST_3, DMA_TX_TRANSFER, dma_id_11 , NULL};
+const dma_descr_t dma_SDIO_0_TX= { DMA2_Channel5, DMA_REQUEST_7, DMA_TX_TRANSFER, dma_id_11 , &dma_init_struct_sdio};
+const dma_descr_t dma_I2C_1_RX = { DMA2_Channel6, DMA_REQUEST_5, DMA_RX_TRANSFER, dma_id_12 , &dma_init_struct_spi_i2c};
+const dma_descr_t dma_I2C_1_TX = { DMA2_Channel7, DMA_REQUEST_5, DMA_TX_TRANSFER, dma_id_13 , &dma_init_struct_spi_i2c}; */
+
+#define NSTREAMS_PER_CONTROLLER ((1 << NSTREAMS_PER_CONTROLLER_LOG2)-1)
+#define NSTREAM                 (NCONTROLLERS * NSTREAMS_PER_CONTROLLER)
+static const uint8_t dma_irqn[NSTREAM] = {
+    DMA1_Channel1_IRQn,
+    DMA1_Channel2_IRQn,
+    DMA1_Channel3_IRQn,
+    DMA1_Channel4_IRQn,
+    DMA1_Channel5_IRQn,
+    DMA1_Channel6_IRQn,
+    DMA1_Channel7_IRQn,
+    DMA2_Channel1_IRQn,
+    DMA2_Channel2_IRQn,
+    DMA2_Channel3_IRQn,
+    DMA2_Channel4_IRQn,
+    DMA2_Channel5_IRQn,
+    DMA2_Channel6_IRQn,
+    DMA2_Channel7_IRQn,
 };
+#endif
 
 static DMA_HandleTypeDef *dma_handle[NSTREAM] = {NULL};
-static uint8_t  dma_last_channel[NSTREAM];
+static uint8_t  dma_last_subidx[NSTREAM];
 static volatile uint32_t dma_enable_mask = 0;
 
-volatile dma_idle_count_t dma_idle;
-
-#define DMA1_ENABLE_MASK    0x00ff  // Bits in dma_enable_mask corresponfing to DMA1
-#define DMA2_ENABLE_MASK    0xff00  // Bits in dma_enable_mask corresponding to DMA2
-#define DMA_INVALID_CHANNEL 0xff    // Value stored in dma_last_channel which means invalid
-
+#if defined(MCU_SERIES_F4) || defined(MCU_SERIES_F7)
 #define DMA_CHANNEL_AS_UINT8(dma_channel)   (((dma_channel) & DMA_SxCR_CHSEL) >> 24)
```

Comment:
Thanks. But this macro is anyway wrong. In the data sheet the channel selection bits start a position 25 so the shift should be 25. 

The macro should return the Stream number on f4/f7 which is in the range 0..7 and the channel number on l4 which is in the range 1..7. If we keep it simple we can use for both cases the macro
`#define DMA_CHANNEL_AS_UINT8(dma_descr)    ((dma_descr->id) % NSTREAMS_PER_CONTROLLER)`
If  would like to really have the values 1..7 instead of 0..6 for L4 we have to adapt the Macro accordingly.

I make the simple modification and re-push the stuff.